### PR TITLE
Agent reliability / Malformed data

### PR DESCRIPTION
**Pull request by niroyb**
Multiple customers have reported reliability issues when forwarding data (i.e. syslog) to a Linux agent. We have identified (and fixed) some instances where encoding of these forwarded items fail to upload. However, the output plugin will buffer forever when a malformed data item cannot be processed.
=> extend output plugin to drop invalid data and continue processing.
=> Do not retry if uncertain to fail again
@Microsoft/omsagent-devs

### CODE REVIEW

File: source/code/plugins/oms_common.rb [hunk 1]
```diff
@@ -144,23 +149,32 @@ def start_request(req, secure_http, ignore404 = false)
...
-          if res
-            if ignore404 and res.code == "404"
-              return ''
-            end
+          if ignore404 and res.code == "404"
+            return ''
+          end
 
-            res_summary = "(#{res.code} #{res.message} #{res.body})"
+          res_summary = "(#{res.class.name} #{res.code} #{res.message} #{res.body})"
```
Comment (robbiezhang) [inline source/code/plugins/oms_common.rb]: can you add separator between the 4 parts?

File: source/code/plugins/out_oms.rb [hunk 2]
```diff
@@ -95,6 +95,15 @@ def write(chunk)
       unmergable_records.each { |tag, record|
         handle_record(tag, record)
       }
+    rescue OMS::RetryRequestException => e
+      @log.debug "Encountered retryable exception. Will retry sending data later. #{e}"
+      # Re-raise the exception to inform the fluentd engine we want to retry sending this chunk of data later.
```
Comment (robbiezhang) [inline source/code/plugins/out_oms.rb]: is it possible to set an exponential retry interval?

**Comment by robbiezhang** [comment]
Signoff

File: source/code/plugins/out_oms.rb (hunk 2 above)
Comment (niroyb) [inline source/code/plugins/out_oms.rb]: It's already handled by the engine and is configurable. http://docs.fluentd.org/articles/output-plugin-overview#retrywait-maxretrywait

**Pull request merged**